mobly/controllers/android_device.py

```diff
@@ -207,15 +217,16 @@ def get_instances_with_configs(configs):
             raise Error(
                 "Required value 'serial' is missing in AndroidDevice config %s."
                 % c)
-        is_required = c.pop("required", True)
+        is_required = c.get(KEY_DEVICE_REQUIRED, True)
         try:
             ad = AndroidDevice(serial)
             ad.load_config(c)
-        except:
-            if not is_required:
-                logging.warning("Failed to initialize Android device %s", serial)
-                continue
-            raise
+        except Exception as e:
+            if is_required:
+                raise
+            logging.warning("Failed to initialize optional device %s: %s",
+                            serial, e)
+            continue
```

Comment:
For the current use case requested by users, they just want to skip these devices and don't care about the indexing. Because they are simply applying the same logic to a list of devices with no assumption on order etc.

I don't think appending None is a good idea because that says nothing about the identity of the missing object, and adds complications for logics that iterate through the device list.

Since one has to use the json format, the user should not rely on index order to obtain device object should it be necessary to id certain devices in the test bed. They should figure out the identity via android_device.get_device and filter by 'label' or other field.

So if a logic does need to know the object's role in the test bed, it can be properly managed. Appending None doesn't really benefit anything.